File: Programmers/Lv2/쿼드압축 후 개수 세기.py
def solution(arr):
    answer = [0, 0]
    n = len(arr)

    def Quad(a, b, length):
        start = arr[a][b]
        # 시작값 확인
        for i in range(a, a + length):
            for j in range(b, b + length):
                if arr[i][j] != start:  # 다른 수 있으면 재귀적으로
                    length //= 2
                    Quad(a, b, length)
                    Quad(a, b + length, length)
                    Quad(a + length, b, length)
                    Quad(a + length, b + length, length)
                    return
                # 다른 수 없으면 재귀 끝, 해당 tmp 배열의 시작 수 인덱스 증가
        answer[start] += 1

    Quad(0, 0, n)

    return answer

File: Programmers/Lv2/쿼드압축 후 개수 세기.js
function solution(arr) {
  var answer = [0, 0];
  let n = arr.length;

  const check = (x, y, n) => {
    const start = arr[x][y];
    for (let i = x; i < x + n; i++) {
      for (let j = y; j < y + n; j++) {
        if (arr[i][j] !== start) {
          return -1;
        }
      }
    }
    return start;
  };

  const divideAndCount = (x, y, n) => {
    if (n === 1) {
      // 기저 조건: 최소 영역에 도달한 경우 처리
      answer[arr[x][y]]++;
      return;
    }

    const tmp = check(x, y, n);

    if (tmp !== -1) {
      answer[tmp]++;
    } else {
      let half = n / 2;
      divideAndCount(x, y, half);
      divideAndCount(x, y + half, half);
      divideAndCount(x + half, y, half);
      divideAndCount(x + half, y + half, half);
    }
  };

  divideAndCount(0, 0, n);
  return answer;
}
